### ISSUE TITLE
TrainLoraNode: memory-dependent attention chunking can change between forward and gradient-checkpoint recomputation

### ISSUE THREAD

**Issue by pozor** [comfyanonymous/ComfyUI#15845]
### Custom Node Testing

- [x] I have tried disabling custom nodes and the issue persists (see [how to disable custom nodes](https://docs.comfy.org/troubleshooting/custom-node-issues#step-1%3A-test-with-all-custom-nodes-disabled) if you need help)

### Expected Behavior

Gradient-checkpoint recomputation should execute a computation structurally
compatible with the original forward, independently of transient memory
pressure. Training should not fail depending on how much memory happened to be
free at the moment each attention call was made.

### Actual Behavior

`TrainLoraNode` fails during backward with
`torch.utils.checkpoint.CheckpointError` because the attention implementation
chooses its chunk size from the amount of free memory available at runtime.
The original forward and the gradient-checkpoint recomputation can observe
different free-memory values and therefore split the same attention sequence
differently.

**Root cause, isolated by a paired A/B run** (details under Steps to Reproduce):
in `comfy/ldm/modules/attention.py`, the sub-quadratic attention path calls
`model_management.get_free_memory(...)` and then selects `query_chunk_size`
from the list:

```
[4096, 2048, 1024, 512, 256]
```

by finding the largest `x` for which
`mem_free_total / (batch_x_heads * bytes_per_token * x * 4.0) >= k_tokens`.

For the failing call in my reproduction:

- `batch_x_heads = 30`
- dtype bf16, so `bytes_per_token = 2`
- sequence length 2176 tokens

Under that heuristic, a 4096 chunk requires roughly 2.14 GB free and a 2048
chunk roughly 1.07 GB. The forward and the recomputation land on opposite sides
of that threshold:

```
forward:         query_chunk_size = 4096  -> one effective 2176-token chunk
recomputation:   query_chunk_size = 2048  -> 2048 + 128
```

so the first differing saved tensor is `[30, 2176, 128] -> [30, 2048, 128]`,
and every subsequent saved tensor is displaced in the checkpoint sequence.
That displacement is why the exception reports dozens of downstream metadata
mismatches rather than a single one.

Note that `2048` is not an arbitrary number: it is literally one of the
constants in the list above.

This makes runtime free memory part of the computational structure inside a
function that is later recomputed by gradient checkpointing.

**Second isolated variant.** With `--use-split-cross-attention` I observe a
different deterministic mismatch on the same 2176-token sequence, which becomes
1088 during recomputation. That attention path also makes its splitting
decision dynamically from available memory.

I consider the memory-dependent attention splitting root cause isolated for
these two variants only. I have observed other `CheckpointError` variants on
this node which I have **not** isolated, and I am not claiming they share this
cause.

### Steps to Reproduce

Two runs, identical except for the attention backend.

**Common configuration** (Z-Image Base bf16, MPS):

| setting | value |
|---|---|
| dataset | 32 image/text pairs |
| `ImageScaleToTotalPixels` | `megapixels=0.5`, `resolution_steps=16` |
| `TrainLoraNode` | `steps=1`, `seed=0`, `batch_size=1`, rank 16, AdamW, MSE |
| | `gradient_checkpointing=true`, `checkpoint_depth=2` |
| | `offloading=false`, `bucket_mode=true`, `training_dtype=bf16` |

Resolution buckets produced: `110x74:15`, `74x110:11`, `124x66:1`,
`100x82:1`, `154x52:1`, `84x98:1`, `96x84:1`, `118x70:1`.
Both runs deterministically selected bucket 3.

**Run A — default attention.** Launch args `--enable-manager`.
Startup log line: `Using sub quadratic optimization for attention`.
Result: **fails** at step 0 inside `bwd_loss.backward()`.

**Run B — PyTorch attention.** Launch args
`--enable-manager --use-pytorch-cross-attention`.
Startup log line: `Using pytorch attention`.
Result: **passes**, forward and backward both complete.

The two runs were performed seven minutes apart on the same machine and
session. `offloading` was `false` in both, so the custom
`OffloadCheckpointFunction` path in `nodes_train.py` is not required to trigger
this failure.

There was ~14.2 GB reported free on MPS when the exception was raised, so this
is not an out-of-memory failure.

**On reproducer size:** I have not reduced the dataset below 32 images. The
failing branch explicitly depends on runtime free memory, so shrinking the
dataset may change the threshold crossing itself and make the failure disappear
for reasons unrelated to the defect. The image content is irrelevant; the
memory state produced by the pipeline is not.

### Debug Logs

```powershell
Run A (fails):


Using sub quadratic optimization for attention
...
Loaded 32 images from .../input/<dataset>
Resolution bucket (110x74): 15 samples
Resolution bucket (74x110): 11 samples
Resolution bucket (124x66): 1 samples
Resolution bucket (100x82): 1 samples
Resolution bucket (154x52): 1 samples
Resolution bucket (84x98): 1 samples
Resolution bucket (96x84): 1 samples
Resolution bucket (118x70): 1 samples
Created 8 resolution buckets from 32 samples
model weight dtype torch.bfloat16, manual cast: None
model_type FLOW
Gradient checkpointing: patching 40 modules at depth 2
Requested to load Lumina2
loaded completely;  11825.14 MB loaded, full load: True
Training LoRA:   0%|          | 0/1 [00:02<?, ?it/s]
!!! Exception during processing !!! torch.utils.checkpoint: Recomputed values
for the following tensors have different metadata than during the forward pass.
tensor at position 37:
saved metadata:      {'shape': torch.Size([30, 2176, 128]), 'dtype': torch.bfloat16, 'device': device(type='mps', index=0)}
recomputed metadata: {'shape': torch.Size([30, 2048, 128]), 'dtype': torch.bfloat16, 'device': device(type='mps', index=0)}
tensor at position 39:
saved metadata:      {'shape': torch.Size([30, 2176, 2176]), 'dtype': torch.float32, 'device': device(type='mps', index=0)}
recomputed metadata: {'shape': torch.Size([30, 2048, 2176]), 'dtype': torch.float32, 'device': device(type='mps', index=0)}
tensor at position 41:
saved metadata:      {'shape': torch.Size([30, 2176, 2176]), 'dtype': torch.bfloat16, 'device': device(type='mps', index=0)}
recomputed metadata: {'shape': torch.Size([30, 2048, 2176]), 'dtype': torch.bfloat16, 'device': device(type='mps', index=0)}
[... ~45 further displaced tensors omitted ...]


Stack trace (abridged):


  File "comfy_extras/nodes_train.py", line 1102, in _run_training_loop
    guider.sample(...)
  File "comfy_extras/nodes_train.py", line 218, in outer_sample
    output = self.inner_sample(...)
  File "comfy_extras/nodes_train.py", line 518, in sample
    self._train_step_bucket_mode(...)
  File "comfy_extras/nodes_train.py", line 414, in _train_step_bucket_mode
    loss = self.fwd_bwd(...)
  File "comfy_extras/nodes_train.py", line 381, in fwd_bwd
    bwd_loss.backward()
  File "torch/utils/checkpoint.py", line 1177, in unpack_hook
    frame.check_recomputed_tensors_match(gid)
  File "torch/utils/checkpoint.py", line 921, in check_recomputed_tensors_match
    raise CheckpointError(...)


Run B (passes), same configuration, only the launch arg changed:


Using pytorch attention
...
Created 8 resolution buckets from 32 samples
Gradient checkpointing: patching 40 modules at depth 2
loaded completely;  11825.14 MB loaded, full load: True
Training LoRA: 100%|##########| 1/1 [00:11<00:00, 11.14s/it, loss=3.1607, bucket=3]
```

### Other

**Environment**

- ComfyUI commit `4da9e2dbead52fc1e68beae33fe3d7ad63b63241`, base tag `v0.33.3`
- `comfy_extras/nodes_train.py` as of commit `d0fec2ef` (2026-07-21)
- PyTorch 2.10.0, Python 3.13.12
- macOS 26.6.2, Apple Silicon M5 Pro, 24 GB unified memory, device MPS
- Model: Z-Image Base, bf16

This occurs on stock ComfyUI. No local patch is required to reproduce it.

**Workaround**

`--use-pytorch-cross-attention` avoids the memory-dependent chunking path on
this setup. I have since run multi-hour trainings with PyTorch attention
without encountering this `CheckpointError`.

**Possible directions for a fix** — I have not implemented or tested any of
these, so I am offering them as directions rather than a proposed patch:

- cache the chunking decision made during the original forward and reuse it
  during recomputation;
- derive `query_chunk_size` from static quantities (sequence length, head
  count, dtype) rather than from instantaneous free memory;
- have gradient checkpointing pin the relevant decision for the duration of a
  checkpointed region.

**Possibly related**

- #14837 — silent MPS attention corruption above ~2^31 elements. Different
  failure mode (my attention matrix is ~1.4e8 elements, far below that
  threshold), but it makes the same structural observation: the chunking logic
  in these implementations is "free-memory-based, designed to avoid running out
  of memory", not to guarantee anything else.
- #15804 — MPS + bf16 NaN in `sub_quadratic_attention`. Different symptom, same
  file and platform.
- #11124, #10940 — other `TrainLoraNode` issues, unrelated causes.
